cmd/snap-bootstrap/initramfs_install_handler.go

```diff
@@ -0,0 +1,418 @@
+	memDB, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
+		Backstore: asserts.NewMemoryBackstore(),
+		Trusted:   trusted,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	commitTo = func(b *asserts.Batch) error {
+		return b.CommitToAndObserve(memDB, commitObserve, nil)
+	}
+
+	return memDB, commitTo, nil
+}
+
+func loadDeviceSeed(systemLabel string) (deviceSeed seed.Seed, err error) {
+	deviceSeed, err = seed.Open(boot.InitramfsUbuntuSeedDir, systemLabel)
+	if err != nil {
+		return nil, err
+	}
+	n := runtime.NumCPU()
+	if n > 1 {
+		deviceSeed.SetParallelism(n)
+	}
+	return deviceSeed, err
+}
+
+func loadSeedAssertions(deviceSeed seed.Seed) (*asserts.Database, error) {
+	start := time.Now()
+	db, commitTo, err := newMemAssertionsDB(nil)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open temp database: %v\n", err)
+	}
+	if err = deviceSeed.LoadAssertions(db, commitTo); err != nil {
+		return nil, fmt.Errorf("Failed to load assertions: %v\n", err)
+	}
+	logger.Noticef("Seed assertions loaded in %s\n", time.Since(start))
+	return db, nil
+}
+
+func loadSeedEssentialMeta(deviceSeed seed.Seed, sysLabel string) error {
+	perfTimings := timings.New(nil)
+	// measure meta loading
+	start := time.Now()
+	logger.Noticef("Starting to load Seed Meta\n")
+	snap.SanitizePlugsSlots = builtin.SanitizePlugsSlots
+
+	if err := deviceSeed.LoadEssentialMeta([]snap.Type{snap.TypeGadget, snap.TypeBase, snap.TypeKernel}, perfTimings); err != nil {
+		return fmt.Errorf("Failed to load essential seed meta: %v\n", err)
+	}
+
+	logger.Noticef("Seed meta loaded in %s\n", time.Since(start))
+	return nil
+}
+
+// helper to check if slice contains
+// TODO: remove when moving to go 1.18
+func contains(s []snap.Type, e snap.Type) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func mountEssentials(deviceSeed seed.Seed, types []snap.Type) error {
+	// iterate over essential snaps and mount gadget and kernel
+	sd := systemd.New(systemd.SystemMode, progress.Null)
+	for _, s := range deviceSeed.EssentialSnaps() {
+		if contains(types, s.EssentialType) {
+			mountPoint := filepath.Join(boot.InitramfsRunMntDir, snapTypeToMountDir[s.EssentialType])
+			if err := doSystemdMount(s.Path, mountPoint, mountReadOnlyOptions); err != nil {
+				return err
+			}
+			activeMounts = append(activeMounts, mountInfo{
+				mountPoint: mountPoint,
+				unmountFnc: func() error {
+					return sd.Umount(mountPoint)
+				},
+				essential: true,
+			})
+		}
+	}
+	return nil
+}
+
+func umountEssentials() error {
+	for i := len(activeMounts) - 1; i >= 0; i-- {
+		if activeMounts[i].essential {
+			err := doUnmount(&activeMounts[i])
+			if err != nil {
+				return err
+			}
+			// remove from array
+			activeMounts = append(activeMounts[:i], activeMounts[i+1:]...)
+		}
+	}
+	return nil
+}
+
+func doRunInstall(deviceSeed seed.Seed, recoverySystemLabel string) error {
```

Comment:
we need to find a way to share even more of this with install handler code

Reply:
Ultimately I think we should move the install handler to bootstrap.